### PR TITLE
Upgrade to lsp-1.0

### PR DESCRIPTION
**Pull request by wz1000**
This involved a signficant rework of all LSP interaction and `hls-plugin-api`

TODO

- [x] Release new versions of `lsp`,`lsp-test`
- [x] Make `lsp-test` automatically respond to "progress create" and "apply edit" requests.
- [x] Port GHCIDE tests
- [x] Port HLS plugins
- [x] Restore `executeCommand`
- [x] Port HLS tests
- [x] Restore tracing/telemetry
- [x] Fix warnings/hlints

### CODE REVIEW

File: ghcide/src/Development/IDE/Core/Service.hs [hunk 1]
```diff
@@ -37,26 +37,18 @@ import Control.Monad
...
 
 -- | Initialise the Compiler Service.
-initialise :: LSP.ClientCapabilities
-           -> Rules ()
-           -> IO LSP.LspId
-           -> (LSP.FromServerMessage -> IO ())
-           -> WithProgressFunc
-           -> WithIndefiniteProgressFunc
+initialise :: Rules ()
+           -> Maybe (LSP.LanguageContextEnv Config)
```
Comment (bubba) [inline ghcide/src/Development/IDE/Core/Service.hs:41]: Hiding the LSP functionality into LanguageContextEnv is exactly what I had in mind 👍

File: ghcide/src/Development/IDE/Core/Shake.hs [hunk 2]
```diff
@@ -742,6 +739,18 @@ instantiateDelayedAction (DelayedAction _ s p a) = do
       d' = DelayedAction (Just u) s p a'
   return (b, d')
 
+mRunLspT :: Applicative m => Maybe (LSP.LanguageContextEnv c ) -> LSP.LspT c m () -> m ()
+mRunLspT (Just lspEnv) f = LSP.runLspT lspEnv f
+mRunLspT Nothing _ = pure ()
```
Comment (bubba) [inline ghcide/src/Development/IDE/Core/Shake.hs:744]: So when invoking ghcide from the cli the `LanguageContextEnv` should be null right?

**Review by bubba — COMMENTED** [review]
Just did a quick scan, but so far makes a lot of sense. Did ghcide use a reactor style channel beforehand?

File: ghcide/src/Development/IDE/Core/Shake.hs (hunk 2 above)
Comment (wz1000) [reply ghcide/src/Development/IDE/Core/Shake.hs:744]: Yeah

**Comment by wz1000** [comment]
The major changes introduced by this PR are now pretty much ready, here is a summary:

# lsp-1.0

`lsp-1.0` brings many improvments to the LSP library, with a much nicer interface and type-safety as well as lots of bug fixes.

For a summary of the changes, see the release notes on https://github.com/alanz/lsp/releases/tag/1.0.0.0

For an example of the concrete differences introduced due to this API, consider the `request` function from `lsp-test`.

In the old version, this is its type:

```haskell
request :: (ToJSON params, FromJSON a) => ClientMethod -> params -> Session (ResponseMessage a) 
```

To use this method, the user had to guess the correct type of `params` and `a`(the shape of the response) themselves, and ensure that it matched up with the `ClientMethod` in the first parameter.

Using `lsp-1.0`, we can give the following type to `request`:

```haskell
request :: SClientMethod m -> MessageParams m -> Session (ResponseMessage m)
```

[`MessageParams`](https://hackage.haskell.org/package/lsp-types-1.0.0.1/docs/Language-LSP-Types.html#t:MessageParams) is a type family that correctly uses the
type of the parameters that need to be supplied to the request, depending on the
`ClientMethod` that was specified as the first argument

`ResponseMessage` is also parameterized over the LSP method, and is guaranteed
contain a response of the correct "shape".

# Changes in `hls-plugin-api`

Overall, the changes serve to unify the handling 

The `PluginDescriptor` record doesn't need to have a seperate field
for each type of request that can be responded to by plugins.
Instead, we use a `PluginHandlers` type that contains a ["dependent
map"](https://hackage.haskell.org/package/dependent-map) from LSP request
methods to the handlers for those methods.

To add a handler for a message type `m` into this map, it must implement the `PluginMethod` typeclass. This typeclass specifies two things:

1. How do we combine multiple responses from different plugins into a single response?
2. Given a plugin id, how do we check if this plugin is enabled for this method type in the config

To extend 'hls-plugin-api' to support a new kind of request, all that is needed is to
supply an instance for `PluginMethod` for the new request method.

For instance, to support "go to definition" plugins, all we need to do is supply

```haskell
instance PluginMethod TextDocumentDefinition where
  ...
```

The `combineResponses` method of this class now implements much of the logic that previously lived in `ghcide` as part of `Development.IDE.Plugin.HLS` to combine multiple plugin responses.

To define plugin handlers, plugins can use the monoid instance of `PluginHandlers` along with `mkPluginHandler`:

```haskell
mkPluginHandler
  :: PluginMethod m
  => SClientMethod m
  -> PluginMethodHandler ideState m
  -> PluginHandlers ideState

type PluginMethodHandler a m = a -> PluginId -> MessageParams m -> LspM Config (Either ResponseError (ResponseResult m))
```

An example: 

```haskell
descriptor :: PluginId -> PluginDescriptor IdeState
descriptor plId = (defaultPluginDescriptor plId)
  { pluginRules = exampleRules
  , pluginCommands = [PluginCommand "codelens.todo" "example adding" addTodoCmd]
  , pluginHandlers = mkPluginHandler STextDocumentCodeAction     codeAction
                  <> mkPluginHandler STextDocumentCodeLens       codeLens
                  <> mkPluginHandler STextDocumentHover          hover
                  <> mkPluginHandler STextDocumentDocumentSymbol symbols
                  <> mkPluginHandler STextDocumentCompletion     completion
  }

```


# Changes in ghcide

Most of the code that lived in ghcide for interacting with the client had to be re-written, along with changes to the `ghcide` plugin API.

In particular, the `WithMessage` datatype was removed, and we can add handlers for different LSP methods using the `Handler` type and its
`Monoid` instance: https://hackage.haskell.org/package/lsp-1.0.0.1/docs/Language-LSP-Server.html#t:Handlers

We alias the LSP `requestHandler` and `notificationHandler` functions to provide
new versions that don't block the `lsp` message loop, and simply forward the
computation to a channel, where it will be picked up by a ghcide worker thread,
which can decide to execute the computation synchronously (for notifications),
asynchronously (for requests) or not at all (for cancelled requests).

```haskell
requestHandler
  :: forall (m :: Method FromClient Request) c.
     SMethod m
  -> (IdeState -> MessageParams m -> LspM c (Either ResponseError (ResponseResult m)))
  -> Handlers (ServerM c)

notificationHandler
  :: forall (m :: Method FromClient Notification) c.
     SMethod m
  -> (IdeState -> MessageParams m -> LspM c ())
  -> Handlers (ServerM c)
```

Again, these handlers can be composed using `(<>)` or `mconcat`

# Code Actions

Code actions used to return a `Maybe Value` response, as well as a
`WorkspaceApplyEdit` that would be send to the client as a new request. However,
now since all the handlers live inside the `LspM` monad, they can simply use
[`sendRequest`](https://hackage.haskell.org/package/lsp-1.0.0.1/docs/Language-LSP-Server.html#v:sendRequest) to send the apply edit request themselves.

This also allows them to wait for the response to this request, so that the result
of the `ExecuteCommandRequest` can depend on whether the client successfully applied the edit.
However, right now we just ignore the response to preserve the original behaviour.

**Comment by wz1000** [comment]
While there is still some work to be done on this, all the major changes are in, so I would appreciate reviews on the general API and direction of the PR

**Comment by wz1000** [comment]
Also, if anyone wants to help with porting the HLS testsuite, please feel free to reach out. I believe it would be a good exercise  to familiarize yourself with the new `lsp` API :stuck_out_tongue:

**Comment by wz1000** [comment]
Another (maybe helpful) tip for people reviewing the changes, especially to the notifications: ignore white-space changes, that makes a diff a lot less scary

File: ghcide/exe/Main.hs [hunk 3]
```diff
@@ -144,15 +140,16 @@ runIde Arguments{..} hiedb hiechan = do
                           `catchAny` (\e -> (hPutStrLn stderr $ "setInitialDynFlags: " ++ displayException e) >> pure Nothing)
 
             sessionLoader <- loadSession $ fromMaybe dir rootPath
-            config <- fromMaybe def <$> getConfig
+            let config = maybe def id <$> (LSP.runLspT env LSP.getConfig)
```
Comment (Ailrun) [inline ghcide/exe/Main.hs:143]: Out of curiosity: is there a reason for replacing `fromMaybe` with `maybe`?

File: ghcide/src/Development/IDE/Core/Compile.hs [hunk 4]
```diff
@@ -532,45 +532,47 @@ indexHieFile se mod_summary srcPath hash hf = atomically $ do
...
+                Left _err -> pure (Nothing,Nothing)
+                Right _ -> do
+                  LSP.sendNotification LSP.SProgress $ LSP.ProgressParams u $
+                    LSP.Begin $ LSP.WorkDoneProgressBeginParams
+                      { _title = "Indexing references from:"
+                      , _cancellable = Nothing
+                      , _message = Nothing
+                      , _percentage = Nothing
+                      }
+                  pure (Just u, Just u)
```
Comment (Ailrun) [inline ghcide/src/Development/IDE/Core/Compile.hs:557]: It looks like all the branches uses `pure (x, x)`. How about extracting `dupe` from these to make it clear that we are returning the value of `Var`?

File: ghcide/src/Development/IDE/Core/FileStore.hs [hunk 5]
```diff
@@ -77,9 +78,9 @@ makeVFSHandle = do
                     Just content -> Map.insert uri (VirtualFile nextVersion 0 (Rope.fromText content)) vfs
         }
 
-makeLSPVFSHandle :: LspFuncs c -> VFSHandle
-makeLSPVFSHandle lspFuncs = VFSHandle
-    { getVirtualFile = getVirtualFileFunc lspFuncs
+makeLSPVFSHandle :: LanguageContextEnv c -> VFSHandle
+makeLSPVFSHandle lspEnv = VFSHandle
+    { getVirtualFile = \uri -> runLspT lspEnv $ LSP.getVirtualFile uri
```
Comment (Ailrun) [inline ghcide/src/Development/IDE/Core/FileStore.hs:83]: Do we need this explicit eta expansion?

**Review by pepeiborra — APPROVED** [review]
As discussed, let's land this ASAP and deal with bugs/issues later. Dependencies need to be all in Hackage for that though

I plan to build all my future work on top of this branch  so please avoid rewriting the history (just squash merge at the end)

**Comment by alanz** [comment]
FWIW, I have been running with it for the last few days, and seems solid. Plus my tree-sitter version is based on it, and I have not seen anything dire while investigating behaviour for that.

**Pull request merged**